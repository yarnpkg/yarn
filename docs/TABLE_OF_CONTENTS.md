#Table of Contents

This is the documentation for the Yarn package manager, here you can find a simple but complete breakdown of how to install and use Yarn as well as how it operates alongside NPM on your system. 

##Installation

1. [Start Installation](installation/Installation.md)
2. [macOS Install](installation/macosInstall.md)
3. [Linux Install](installation/linuxInstall.md)
4. [Windows Install](installation/windowsInstall.md)

##Usage

1. [Basic Usage](usage/basicUsage.md)
2. [Creating a New Project](usage/creatingNewProject.md)
3. [Managing Dependencies](usage/managingDependencies.md)
4. [Installing Dependencies](usage/installingDependencies.md)
5. [Version Control](usage/versionControl.md)

##NPM Compatability

1. [Migrating From NPM](usage/migratingFromNPM.md)