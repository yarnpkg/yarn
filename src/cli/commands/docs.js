/* @flow */

import opn from 'opn';
import type {Reporter} from '../../reporters/index.js';
import type Config from '../../config.js';
import handlePackageName from '../../util/handle-package-name.js';
import parsePackageName from '../../util/parse-package-name.js';

function openPackageDocs(
  name: string,
  packageInfo,
  app?: string): Promise<string> {
  const options = {};
  let url = packageInfo.homepage;
  if (!url) {
    url = `https://www.npmjs.org/package/${name}`;
  }
  if (app) {
    options.app = app;
  }
  opn(url, options, false);
  return Promise.resolve(url);
}

export function setFlags(commander: Object) {
  commander.usage('docs [packages ...] [flags]');
  commander.option(
    '-a, --app <app>',
    'the app that is used to open websites.' +
    ' Defaults: macOS="open", Windows="start", Others="xdg-open"');
}

export async function run(
 config: Config,
 reporter: Reporter,
 flags: Object,
 args: Array<string>,
): Promise<void> {
  const packageNames = args.slice();
  if (packageNames.length === 0) {
    packageNames.push('.');
  }
  for (let i = 0; i < packageNames.length; i++) {
    const packageName = await handlePackageName(packageNames[i], config);
    if (!packageName) {
      reporter.error(reporter.lang('packageHasNoName', packageName));
      return;
    }
    const {name} = parsePackageName(packageName);
    const packageInfo = await config.registries.npm.request(name);
    if (!packageInfo) {
      reporter.error(reporter.lang('infoFail'));
      return;
    } else {
      const url = await openPackageDocs(name, packageInfo, flags.app);
      reporter.log(reporter.lang('openedDocs', name, url));
    }
  }
}
