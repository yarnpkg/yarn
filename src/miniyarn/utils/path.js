import Path from 'path';

export function validate(path) {
  if (Path === Path.posix && isAbsolute(path) && !normalize(path).startsWith(`/`))
    throw new Error(`This system doesn't support drive letters`);

  return true;
}

export function dirname(path) {
  return normalize(Path.win32.dirname(path));
}

export function normalize(path) {
  // We don't want the "./" prefix, because it makes it harder to apply pattern matching

  return Path.win32.normalize(path).replace(/\\/g, `/`).replace(/\/$/, ``).replace(/^\.\//g, ``);
}

export function isAbsolute(path) {
  return Path.win32.isAbsolute(path);
}

export function isRelative(path) {
  return !isAbsolute(path);
}

export function isForward(path) {
  return isRelative(path) && !normalize(path).startsWith(`../`);
}

export function resolve(...paths) {
  return normalize(Path.win32.resolve(...paths));
}

export function relative(fromPath, toPath) {
  return normalize(Path.win32.relative(fromPath, toPath));
}

export function makeExplicit(path) {
  return normalize(path).replace(/^(?!\.{0,2}\/|\.$)/, `./`);
}

export function isPathForSure(maybePath) {
  return isAbsolute(maybePath) || normalize(maybePath).match(/^\.{1,2}\//);
}
