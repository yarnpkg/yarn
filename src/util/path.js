/* @flow */

const userHome = require('./user-home-dir').default;

export function expandPath(path: string): string {
  if (process.platform !== 'win32') {
    path = path.replace(/^\s*~(?=$|\/|\\)/, userHome);
  }

  return path;
}
