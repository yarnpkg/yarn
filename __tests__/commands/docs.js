/* @flow */

jest.mock('opn');

import {run as docs, setFlags} from '../../src/cli/commands/docs.js';
import {BufferReporter} from '../../src/reporters/index.js';
import Config from '../../src/config.js';
import path from 'path';

jasmine.DEFAULT_TIMEOUT_INTERVAL = 90000;

const fixturesLoc = path.join(__dirname, '..', 'fixtures', 'docs');

async function runDocs(
  args: Array<string>,
  flags: Object,
  name: string,
  mockResponse: ?Object,
  checkSteps?: ?(mocks: Object, output: any) => ?Promise<void>,
): Promise<void> {
  // $FlowFixMe: Update our jest flow-type
  jest.clearAllMocks();
  const reporter = new BufferReporter({stdout: null, stdin: null});
  const config = new Config(reporter);
  const cwd = name && path.join(fixturesLoc, name);
  await config.init({cwd});
  const mocks = {
    opn: require('opn'),
    request: jest.fn(() => mockResponse),
  };
  config.registries = {npm:{request: mocks.request}};
  await docs(config, reporter, flags, args);

  if (checkSteps) {
    const buffer = reporter.getBuffer();
    const output = buffer.pop().data;
    await checkSteps(mocks, output);
  }
}

const testData = {
  'without arguments and in directory containing a valid package file':
    [[], {}, 'valid', {}],
  'without arguments and in directory containing a invalid package file':
    [[], {}, 'invalid', {}],
  'with 1 argument and a invalid request':
    [['jest'], {}, 'valid', null],
  'with 1 argument and a valid request':
    [['jest'], {}, 'valid', {name: 'jest'}],
  'with 1 argument, a valid request, and a homepage':
    [['jest'], {}, 'valid', {name: 'jest', homepage: 'http://example.com'}],
  'with 1 argument, an app passed in, a valid request, and a homepage':
    [['jest'], {app: 'Google Chrome'}, 'valid', {name: 'jest', homepage: 'http://example.com'}],
  'with multiple arguments and a valid request':
    [['jest', 'foo'], {}, 'valid', {name: 'jest'}],
};

Object.keys(testData).forEach((testDescription) => {
  const testArgs = testData[testDescription];
  const [args, flags, name, mockResponse] = testArgs;
  test(testDescription, (): Promise<void> => {
    return runDocs(args, flags, name, mockResponse,
      ({opn, request}, output): ?Promise<void> => {
        expect(opn.mock).toMatchSnapshot();
        expect(request.mock).toMatchSnapshot();
        expect(output).toMatchSnapshot();
      },
    );
  });
});

test('setFlags() modifies our commander object', () => {
  // $FlowFixMe: Update our jest flow-type
  const commander = jest.fn();
  // $FlowFixMe: Update our jest flow-type
  commander.usage = jest.fn();
  // $FlowFixMe: Update our jest flow-type
  commander.option = jest.fn();
  setFlags(commander);
  // $FlowFixMe: Update our jest flow-type
  expect(commander.usage.mock).toMatchSnapshot();
  // $FlowFixMe: Update our jest flow-type
  expect(commander.option.mock).toMatchSnapshot();
});
