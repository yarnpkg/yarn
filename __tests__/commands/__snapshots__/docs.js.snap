exports[`test setFlags() modifies our commander object 1`] = `
Object {
  "calls": Array [
    Array [
      "docs [packages ...] [flags]",
    ],
  ],
  "instances": Array [
    [Function],
  ],
}
`;

exports[`test setFlags() modifies our commander object 2`] = `
Object {
  "calls": Array [
    Array [
      "-a, --app <app>",
      "the app that is used to open websites. Defaults: macOS=\"open\", Windows=\"start\", Others=\"xdg-open\"",
    ],
  ],
  "instances": Array [
    [Function],
  ],
}
`;

exports[`test with 1 argument and a invalid request 1`] = `
Object {
  "calls": Array [],
  "instances": Array [],
}
`;

exports[`test with 1 argument and a invalid request 2`] = `
Object {
  "calls": Array [
    Array [
      "jest",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test with 1 argument and a invalid request 3`] = `"Received invalid response from npm."`;

exports[`test with 1 argument and a valid request 1`] = `
Object {
  "calls": Array [
    Array [
      "https://www.npmjs.org/package/jest",
      Object {},
      false,
    ],
  ],
  "instances": Array [
    undefined,
  ],
}
`;

exports[`test with 1 argument and a valid request 2`] = `
Object {
  "calls": Array [
    Array [
      "jest",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test with 1 argument and a valid request 3`] = `"Opening docs for \"jest\" at \"https://www.npmjs.org/package/jest\""`;

exports[`test with 1 argument, a valid request, and a homepage 1`] = `
Object {
  "calls": Array [
    Array [
      "http://example.com",
      Object {},
      false,
    ],
  ],
  "instances": Array [
    undefined,
  ],
}
`;

exports[`test with 1 argument, a valid request, and a homepage 2`] = `
Object {
  "calls": Array [
    Array [
      "jest",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test with 1 argument, a valid request, and a homepage 3`] = `"Opening docs for \"jest\" at \"http://example.com\""`;

exports[`test with 1 argument, an app passed in, a valid request, and a homepage 1`] = `
Object {
  "calls": Array [
    Array [
      "http://example.com",
      Object {
        "app": "Google Chrome",
      },
      false,
    ],
  ],
  "instances": Array [
    undefined,
  ],
}
`;

exports[`test with 1 argument, an app passed in, a valid request, and a homepage 2`] = `
Object {
  "calls": Array [
    Array [
      "jest",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test with 1 argument, an app passed in, a valid request, and a homepage 3`] = `"Opening docs for \"jest\" at \"http://example.com\""`;

exports[`test with multiple arguments and a valid request 1`] = `
Object {
  "calls": Array [
    Array [
      "https://www.npmjs.org/package/jest",
      Object {},
      false,
    ],
    Array [
      "https://www.npmjs.org/package/foo",
      Object {},
      false,
    ],
  ],
  "instances": Array [
    undefined,
    undefined,
  ],
}
`;

exports[`test with multiple arguments and a valid request 2`] = `
Object {
  "calls": Array [
    Array [
      "jest",
    ],
    Array [
      "foo",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test with multiple arguments and a valid request 3`] = `"Opening docs for \"foo\" at \"https://www.npmjs.org/package/foo\""`;

exports[`test without arguments and in directory containing a invalid package file 1`] = `
Object {
  "calls": Array [],
  "instances": Array [],
}
`;

exports[`test without arguments and in directory containing a invalid package file 2`] = `
Object {
  "calls": Array [],
  "instances": Array [],
}
`;

exports[`test without arguments and in directory containing a invalid package file 3`] = `"The current package.json has no \"name\" attribute"`;

exports[`test without arguments and in directory containing a valid package file 1`] = `
Object {
  "calls": Array [
    Array [
      "https://www.npmjs.org/package/test-docs",
      Object {},
      false,
    ],
  ],
  "instances": Array [
    undefined,
  ],
}
`;

exports[`test without arguments and in directory containing a valid package file 2`] = `
Object {
  "calls": Array [
    Array [
      "test-docs",
    ],
  ],
  "instances": Array [
    Object {
      "request": [Function],
    },
  ],
}
`;

exports[`test without arguments and in directory containing a valid package file 3`] = `"Opening docs for \"test-docs\" at \"https://www.npmjs.org/package/test-docs\""`;
